addOns/ascanrulesBeta/src/main/java/org/zaproxy/zap/extension/ascanrulesBeta/XxeScanRule.java

```diff
@@ -241,35 +241,52 @@ public void scan() {
+     * the process is repeated by replacing one attribute at a time, for a fixed number of
+     * attributes depending on the strength of the rule.
      *
      * @param msg new HttpMessage with the same request as the base. This is used to build the
      *     attack payload.
      */
     private void localFileReflectionAttack(HttpMessage msg) {
         String payload = null;
         try {
-            String requestBody = createLfrPayload(msg.getRequestBody().toString());
-            for (int idx = 0; idx < LOCAL_FILE_TARGETS.length; idx++) {
-                String localFile = LOCAL_FILE_TARGETS[idx];
-                payload = MessageFormat.format(requestBody, localFile);
-                msg.setRequestBody(payload);
-                sendAndReceive(msg);
-                String response = msg.getResponseBody().toString();
-                Matcher matcher = LOCAL_FILE_PATTERNS[idx].matcher(response);
-                if (matcher.find()) {
-                    newAlert()
-                            .setConfidence(Alert.CONFIDENCE_MEDIUM)
-                            .setAttack(payload)
-                            .setEvidence(matcher.group())
-                            .setMessage(msg)
-                            .raise();
+            // First replace the values in all the Elements by the Attack Entity
+            String originalRequestBody = msg.getRequestBody().toString();
+            String requestBody = createLfrPayload(originalRequestBody);
+            boolean alertRaised = localFileReflectionRaiseAlert(msg, requestBody, payload);
+            if (isStop()) {
+                return;
+            }
+            // Now if no issue is found yet, then we replace the values one at a time. Do this for a
+            // fixed number of Elements, depending on the strength at which the rule is used.
+            if (!alertRaised) {
+                // Remove original xml header
+                Matcher headerMatcher = xmlHeaderPattern.matcher(originalRequestBody);
+                String headerlessRequestBody = headerMatcher.replaceAll("");
+                int totalTagCount = getTotalTagCount(headerlessRequestBody);
+                int maxValuesChanged = 0;
+
+                if (this.getAttackStrength() == AttackStrength.MEDIUM) {
+                    maxValuesChanged = 72 / LOCAL_FILE_TARGETS.length;
+                } else if (this.getAttackStrength() == AttackStrength.HIGH) {
+                    maxValuesChanged = 144 / LOCAL_FILE_TARGETS.length;
+                } else if (this.getAttackStrength() == AttackStrength.INSANE) {
+                    maxValuesChanged = totalTagCount;
                 }
-                if (isStop()) {
-                    return;
+                Matcher tagMatcher = tagPattern.matcher(headerlessRequestBody);
+                for (int tagIdx = 1;
+                        tagIdx <= Math.min(totalTagCount, maxValuesChanged);
```

Comment:
This can be done once, before the loop.

Reply:
Or better, since the `Matcher` is now kept between tests it does not need to count, can do `tagIdx <= `maxValuesChanged && tagMatcher.find()`.